### ISSUE TITLE
[RFC] Overhaul lattice link configuration

### ISSUE THREAD

**Issue by autodidaddict** [wasmCloud/wasmCloud#342]
## Summary
This is a _Request for Comment_ on the plan to address a number of instabilities and areas of high friction that are part of the developer experience when interacting with link definitions and lattice link configuration.

### Status Quo: Link Definition Configuration
The following section describes the current implementation of link definition configuration. It is worth going through the status quo in this kind of detail because it helps illustrate the source of a number of the distilled problem cases outlined below.

![image](https://github.com/wasmCloud/wasmCloud/assets/13748641/eba7ab13-c4e8-4434-b0d7-01733fb80ac4)

The current flow when originated from `wash` goes through the following steps (reference to "host" below refers to the OTP host as there is no Rust host at this level of functionality yet):

1. Developer issues a `wash link put` (formerly `wash ctl link put`) command.
2. This writes the link definition to the metadata cache NATS Jetstream key value bucket
3. All running wasmCloud hosts within the lattice will receive a message on a subscription with the action type of `key_add`
4. Each host will use that message to populate their internal in-memory cache
5. Each host then publishes a link definition payload on the `wasmbus.rpc.{provider-key}.{link-name}.link.put` (or similar) topic
6. All running, matching capability providers will receive that message and, if it isn't a duplicate, the `link_put` trait function will be invoked by virtue of `wasmbus_rpc`.
7. Each running suitable target provider in turn receives `h` inbound link put messages.

This means that each time a new link definition is written to the key value bucket, `h*p + h` messages will be published where `h` is the number of hosts and `p` is the number of providers. So, for a 3 host environment with two matching providers, `9` messages will be published (this is depicted in the diagram above)

## Rationale
The rationale, first and foremost, is to address the items from the preceding bullet list of distilled problem areas. In general, developer interaction with link definitions, links in general, and even the overall configuration of a lattice is difficult to understand and manage properly.

### Failure Cases
The following is a detailed enumeration of some of the frequently recurring failure cases that we have identified with the status quo of capability providers and their configuration.

#### Partial Zombie
In this scenario, there are multiple running copies of the same provider+link within a given lattice. Each one received the link put message either in real-time or during the startup configuration load. One of these providers encountered a network partition event during the client resource provisioning phase (e.g. creating a Redis connection) and one did not. The link definition is stored in the key value bucket, it is stored in the host memory cache and, depending on how the provider responded to `link_put`, `wasmbus_rpc` could think the link was successfully established (and will therefore ignore subsequent `puts`).

This problem manifests where roughly half of the requests of this provider are successful and the other half fail. In partitioned leaf networks, **all** requests to the provider from actors running on hosts on the same side of a leaf node will fail while the actors "local" to the stable provider will get a 100% success rate. This type of failure is excruciatingly difficult to troubleshoot because every attempt to query link definitions can lead the developer to believe everything is fine. Sadly, the fix is often to "bounce" all of the providers, giving them all a chance to re-establish client resources at startup. If there are no network failures, this can fix the problem and mask the true cause.

#### Full Zombie
This happens more frequently. In the simple base case when we have an actor and a single running instance of a provider (like many of our samples), we can produce a 100% failure rate. A link definition propagates through the system per the above diagram. When `link_put` is called on the capability provider, it fails outright. This could be due to an invalid connection string. If no one was watching the logs at the exact moment of link put, _the system will have no visible sign that the provider is in a corrupt state_. If we use `wash` to query the link definitions, they will all be there (because _**all**_ link definitions make it into the bucket). If we query host inventory, we will see both the actor and the provider. The only sign something is wrong is seeing various errors in the logs when we attempt to invoke the provider sending either to or from the actor.

In either full or partial zombie modes, a capability provider could even respond positively to a health check, further aggravating the situation.

#### Late Failure
Assume that the the initial link definition put (which can occur at provider start or at time of link put) was successful. The actor and provider continue in a steady state for a while and so the application functions properly. At some point, a runtime condition like a panic in the provider occurs. The provider process could continue to run, so the OTP pipe would not detect a "Dead" provider. The provider could then lose its NATS connection or subscriptions, preventing it from responding to RPC invocations. 

#### The Sleeper Failure (Bad Config, Stable Provider)
Despite how rare this sounds, it has happened frequently enough for us to recognize it. In this situation, the link definition put results in the successful creation of a client or of any resources that need to be provisioned. It isn't until an actual call from an actor arrives that we exercise the client, resulting in an error. If the provider is well-behaved, it will capture the error and return an indication that the call failed. We can then continue with a 100% call failure rate without much of an obvious reason why (other than, again, landing in the lucky spot in the logs). If we catch this error and attempt to re-publish the link with the updated data, all of the target providers will ignore it because of `wasmbus_rpc`'s anti-dupe checker. If we delete and re-add the link, the asynchronous nature of both of those operations could theoretically result in one of the aforementioned "zombie" scenarios. This could even result in yet another sleeper failure, further frustrating those attempting to identify the cause of the problem.

### ⚗️Failure Cases Distilled
The following is a distillation and high level description of the problems identified from the preceding failure cases and other evidence we've gathered from troubleshooting and developer feedback:

* Writing a link definition to the cache works all the time, regardless of provider state
* A provider can fail to process a link and provision resources without anyone noticing
* A provider can be in a dead or partially dead state and attempts to query link definitions and host inventory can appear as though all is good
* In the case of multiple instances of the same provider within a lattice, some could be in an error state while others aren't, resulting in (seemingly) unpredictable failures and failure rates.
* 4 Conflicting sources of truth
  * In-memory cache in OTP host
  * In-memory cache in provider (by way of `wasmbus_rpc`'s link cache)
  * Contents of key-value bucket 
  * Periodic publications on `wasmbus.rpc.{provider}.{link}.xx`
  * Unknown number of race conditions that arise from concurrent manipulation of these sources

At any given time, developers have low confidence in their ability to determine the status of a link definition, the status of the link with respect to the health and status of the provider, and any potential drift and conflict in the multiple sources of truth.

## Implementation Plan
The following is a set of discrete changes that we think we could make to dramatically improve the situation. This should increase reliability, predictability, stability, and make huge improvements to the developer experience while we're at it.

### Invert Config Source - Pull not Push
A fundamental property of a link definition is that it is declarative, and it exists at the lattice level, _without regard for whether its corresponding actor or provider are running_. This fact makes it a bit tricky to deal with publishing link advertisement messages and being able to determine failure since the link initialization can take place practically at any time.

By inverting the model, we can avoid this uncertainty entirely. Instead of pushing link definitions to providers and hoping they're around to "hear it", providers will need to _pull_ their configuration when they need it. This has a number of benefits. The first, is that resources required for a link won't actually be consumed until the link is first exercised because the provider would make a cache miss, pull configuration, and then establish the link. Providers can choose to pull their configuration whenever they like.

Providers make a request on `wasmbus.{host-id}.links.get` to get a list of link definitions and their values. The response is filtered according to the request payload, which is a JWT bearer token. The low-level details don't matter much yet, but the JWT will contain information to identify the provider by its key and link name. The reply will be links for all actors bound to that provider+link. This will be a secure call because of the signature on the JWT, and there's no danger of one provider exfiltrating values for another.

The provider communicates with the host that started it because the two will have established a two-way link when the host supplied information at launch via `stdio`. This also ensures that the absolute closest entity is what responds to the provider, avoiding costly network hops.

### Establish a Link Ping
One of the core problems is we don't really know how well established a given link is, or if it's become broken since it was originally created. To counteract that, we propose that we create a "link ping", which will occur on a topic all providers _must_ subscribe to.

Provider instances should subscribe to the following wildcard at the end of their startup sequence: `wasmbus.link.*.{provider key}.{link name}.>`. This will allow the provider to respond to link pings for any actor. For the insatiably curious, this topic structure implies that a single capability provider cannot implement or respond to more than one contract. We're fine with this for a number of reasons beyond the scope of this RFC.

Submit an empty request to the subject `wasmbus.link.{actor key}.{provider key}.{link name}.ping`. Then, collect responses until a timeout occurs, the results will look like the following JSON:

```json
{"host_id": "Nxxx", "status": "healthy", "uptime_ms": 3000 }
{"host_id": "Nyyy", "status": "down", "uptime_ms": 0 }
{"host_id": "Nzzz", "status": "healthy", "uptime_ms": 6000 }
```
This shows 3 capability providers responding to the ping across the lattice. Each identifies its host, the status of the link, and  how long since the link was established. Note that the process cannot identify provider instances that _should have_ responded but didn't. Anything consuming this API will need to use out of band knowledge to make that assessment.

If you want to perform a ping _and_ query each provider for what it thinks its current configuration values are for that link, then replace `ping` in the query topic with `vping`. By using a different topic segment, NATS user credentials can be banned from using this topic for security reasons.

```json
{"host_id": "Nxxx", "status": "healthy", "uptime_ms": 3000, "values": { "port": 8080 }}
```
This link ping support would be added to all existing capability providers. Given the forthcoming deprecation of `wasmbus_rpc`, this ping functionality may be supported by the replacement "provider SDK".

### Support a Link Purge Request
In most cases, such a request shouldn't be necessary. Inverting the control so the providers pull configuration lazily only when initially needed eliminates a large number of potentially false sources of truth. However, if a link definition is deleted or if the data for that link definition is known to be corrupt, then we'll need a way to tell the provider to purge its current cache, which would in turn instruct it to re-pull that data when next needed. `wash` might invoke this when deleting a link from the key value bucket.

Publish an empty request to `wasmbus.link.{actor key}.{provider key}.{link name}.purge` and, if the provider is conforming to this new model, it will purge that link configuration and can decide when to re-pull it. It could decide to pull right away or wait until the next time that link is utilized.

## Risks
There are a number of breaking changes here. Capability providers that conform to this new system would have a number of subscriptions to NATS topics that "old" capability providers would not. These new providers would likely ignore the current link definition advertisement topics that the host would publish on for backwards compatibility. 

Old and new providers would be able to co-exist on the same host (while the old advertisement topic is deprecated), but new providers would _not_ be able to function properly on old hosts. This is fine with us as we would retain backwards compatibility, which is more important than forwards.

This would have little to no impact on `wadm`, but `wadm` could be made even more deterministic if it made use of link ping and purge topics.

**Comment by autodidaddict** [comment]
⚠️ **still in draft**. Saved it to avoid losing all those words if my tab dies ;)

**Comment by connorsmith256** [comment]
This is an excellent overview of the rough edges of the status quo 🙌 

Some thoughts/questions on the proposed implementation:
- Since the source of truth for link defs is the KV bucket, why does the host need to be involved when a provider queries its links? Could the provider not query the bucket directly instead of requesting on `wasmbus.{host-id}.links.get`?
  - Related: if providers interact with the source of truth directly, `wasmbus.link.{actor key}.{provider key}.{link name}.purge` might not be needed either. The provider could subscribe to delete/update events in the bucket and update its internal state in response
- Under the assumption `wasmbus.{host-id}.links.get` is needed, why do we use `host-id` instead of `lattice-id`? The host that started the provider process and set up a subscription on this topic will be connected to the same leaf node as the provider, so NATS will keep this traffic local to the leaf, right?
- Can you go into more detail on what the "status" of a link means? Is this set once per provider instance/host ID when publishing the link def, or could providers update the status after the link was initially published?
  - My $0.02: I think dynamic, provider-instance-specific state for links will be helpful. I'm thinking of cases where the link was valid at the time the link was created, but has since "broken". E.g. a DB connection string which becomes invalid only after a DNS change is applied. However, how and when would this status get set? It shouldn't be coupled to any individual invocation, since there might be an issue with the invocation payload rather than the underlying config. This is related to the top-level status/health check for a provider instance. Should we get rid of this and introduce a per-link status?
- I like the link ping topic. Querying a provider's internal state would be helpful when troubleshooting 👍

**Comment by autodidaddict** [comment]
Thanks for the thoughtful questions! I think I can answer a few of them with one of the requirements/invariants for this system. For security reasons, the host supplies the capability provider with its NATS connection credentials. In production, a provider should **never** be allowed to directly interact with key value buckets. The connection passed to a provider could be a user credential that only has limited access to RPC topics. In addition to the security concerns, there is also some complexity that comes from the use of a JS domain. There are common topologies where in some places the js domain would be required and not in others. While we can account for this, trying to maintain "js domain cohesion" adds a layer of complexity I think we can avoid.

The host ID is used for the provider to communicate _securely_ with the host. The host can provide a nonce in the provider information that a provider must sign in order to communicate with the host "backplane".  This ensures that the provider doesn't need to make a lattice-wide grab for data (performance), doesn't violate multi-tenancy and security rules by talking to KV buckets, and can only make this request of the host that spawned it. There are a bunch of intrusion vectors that this design blocks that we can't block with a lattice-wide query.

Link status is not permanent or recorded anywhere. It doesn't get pushed to a bucket or anything else. The only way to tell the status of a link would be to use the link ping, which will determine a number of different statuses. It can detect when the provider isn't responding at all, and it can detect when the provider gets the ping and reports an error with that link.

I think one thing that I could've cleared up in the original text is that providers are required to not have access to key-value buckets in "real world" deployments.

**Comment by autodidaddict** [comment]
Implicit in the previous comment - giving a provider the ability to read from the links bucket allows it to exfiltrate all configuration data for all other links in the lattice.

**Comment by autodidaddict** [comment]
I think it would be up to the provider. If a provider has a thread (or task) devoted to a particular actor, and that dies, the provider can either set its own internal state, or wait until it gets pinged to answer. The requirement is that we be able to do a ping of a link and get a response from every provider in the lattice managing that link. So, if we're running 3 instances of the Redis provider for a single link, the ping will give us all 3 statuses.

**Comment by thomastaylor312** [comment]
Also concur that this is really great. Just one question I had:

In this new paradigm, how do requests get routed? So let's say we have a provider running on host A and on host B the provider has failed. When an actor makes a request to the provider, will host B be able to say "sorry I can't handle this request" so that host A gets it instead? I also understand that this could be out of scope for this change as at least users can now debug why some of their requests are failing

**Comment by autodidaddict** [comment]
This RFC doesn't cover routing around dead links for rpc. The topic space doesn't actually support that kind of behavior atm. if we want to address that as a new feature I think that is a different scope.

**Comment by thomastaylor312** [comment]
Oh, one more thought: When we implement this, we might want to implement a "best practice" and/or built in version of caching linkdefs in a provider to make it easy on people writing their own provider

**Comment by autodidaddict** [comment]
Yep that's a great idea. Could maybe be a part of the provider sdk if/when that becomes a thing.

**Comment by brooksmtownsend** [comment]
This will be partially addressed by #1389 regarding link configuration but we still need to answer the question of how we address link configuration failures 🤔
